app/Actions/WindowFactory.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+/*
+ * This file is part of Cachet.
+ *
+ * (c) Alt Three Services Limited
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace CachetHQ\Cachet\Integrations\Core;
+
+use CachetHQ\Cachet\Models\TimedAction;
+use Carbon\Carbon;
+
+/**
+ * This is the window factory class.
+ *
+ * @author Graham Campbell <graham@alt-three.com>
+ */
+class WindowFactory
+{
+    /**
+     * Get the current window.
+     *
+     * @param \CachetHQ\Cachet\Models\TimedAction $action
+     *
+     * @throws \CachetHQ\Cachet\Actions\ActionNotStartedException
+     *
+     * @return \CachetHQ\Cachet\Actions\Window
+     */
+    public function current(TimedActionModel $action)
+    {
+        return new Window($this->currentWindowStart($action), $action->window_length);
+    }
+
+    /**
+     * Get the previous window.
+     *
+     * @param \CachetHQ\Cachet\Models\TimedAction $action
+     *
+     * @throws \CachetHQ\Cachet\Actions\ActionNotStartedException|\CachetHQ\Cachet\Actions\ActionNotMaturedException
+     *
+     * @return \CachetHQ\Cachet\Actions\Window
+     */
+    public function previous(TimedActionModel $action)
+    {
+        return new Window($this->previousWindowStart($action), $action->window_length);
+    }
+
+    /**
+     * Get the start time of the current window.
+     *
+     * @param \CachetHQ\Cachet\Models\TimedAction $action
+     *
+     * @throws \CachetHQ\Cachet\Actions\ActionNotStartedException
+     *
+     * @return \Carbon\Carbon
+     */
+    protected function currentWindowStart(TimedActionModel $action)
+    {
+        $now = Carbon::now();
+
+        $diff = $action->start_at->diffInSeconds($now);
+
+        if ($diff >= 0)
+            throw new ActionNotStartedException("The timed action is only due to start in {$diff} seconds");
```

Comment:
Hmm, I prefer the current phrasing. Makes it sound like an error instead of just a statement.